controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java

```diff
@@ -1974,22 +1976,30 @@ private SegmentRange convert(String scope, String stream, Map.Entry<Long, Map.En
         }
     }
 
-    public CompletableFuture<Map<Long, Long>> notifyTxnCommit(final String scope, final String stream,
-                                                   final List<Long> segments, final UUID txnId, long requestId) {
-        Timer timer = new Timer();
+    public CompletableFuture<Map<Long, List<Long>>> mergeTxnSegmentsIntoStreamSegments(final String scope, final String stream,
+                                                                                       final List<Long> segments, final List<UUID> txnIds, long requestId) {
         return Futures.allOfWithResults(segments.stream()
-                .collect(Collectors.toMap(x -> x, x -> notifyTxnCommit(scope, stream, x, txnId, requestId))))
-                .whenComplete((r, e) -> TransactionMetrics.getInstance().commitTransactionSegments(timer.getElapsed()));
+                        .collect(Collectors.toMap(segId -> segId, segId -> mergeTxnSegments(scope, stream, segId, txnIds, requestId))));
     }
 
-    private CompletableFuture<Long> notifyTxnCommit(final String scope, final String stream,
-                                                    final long segmentNumber, final UUID txnId, long requestId) {
-        return TaskStepsRetryHelper.withRetries(() -> segmentHelper.commitTransaction(scope,
+    private CompletableFuture<List<Long>> mergeTxnSegments(final String scope, final String stream,
+                                                                             final long segmentNumber, final List<UUID> txnIds, long requestId) {
+        return TaskStepsRetryHelper.withRetries(() -> segmentHelper.mergeTxnSegments(scope,
                 stream,
                 segmentNumber,
                 segmentNumber,
-                txnId,
-                this.retrieveDelegationToken(), requestId), executor);
+                txnIds,
+                this.retrieveDelegationToken(), requestId).exceptionally(ex -> {
+                    if (ex instanceof WireCommandFailedException) {
+                        WireCommandFailedException wcfe = (WireCommandFailedException) ex;
+                        if (WireCommandFailedException.Reason.SegmentDoesNotExist.equals(wcfe.getReason())) {
+                            final String qualifiedSegmentName = getQualifiedStreamSegmentName(scope, stream, segmentNumber);
```

Comment:
Should'nt we fetch the offset from the target segment and return the value here?

Reply:
This exception is thrown by PravegaRequestProcessor only when TargetSegment(StreamSegment) does not exist. If the source segment does not exist we're already doing this.